sdks/python/src/opik/integrations/dspy/callback.py

```diff
@@ -1,86 +1,74 @@
-from contextvars import ContextVar, Token
 from typing import Any, Dict, Optional, Union
 
 import dspy
-from dspy.utils.callback import BaseCallback
+from dspy.utils import callback as dspy_callback
 
-from opik import opik_context, types
-from opik.api_objects import helpers, span, trace
-from opik.api_objects.opik_client import get_client_cached
+from opik import types, opik_context, context_storage
+from opik.api_objects import helpers, span, trace, opik_client
 from opik.decorator import error_info_collector
 
 ContextType = Union[span.SpanData, trace.TraceData]
 
 
-class OpikCallback(BaseCallback):
+class OpikCallback(dspy_callback.BaseCallback):
     def __init__(
         self,
         project_name: Optional[str] = None,
     ):
         self._map_call_id_to_span_data: Dict[str, span.SpanData] = {}
         self._map_call_id_to_trace_data: Dict[str, trace.TraceData] = {}
-        self._map_span_id_or_trace_id_to_token: Dict[str, Token] = {}
 
         self._origins_metadata = {"created_from": "dspy"}
 
-        self._current_callback_context: ContextVar[Optional[ContextType]] = ContextVar(
-            "opik_context", default=None
-        )
+        self._context_storage = context_storage.OpikContextStorage()
 
         self._project_name = project_name
 
-        self._opik_client = get_client_cached()
+        self._opik_client = opik_client.get_client_cached()
 
     def on_module_start(
         self,
         call_id: str,
         instance: Any,
         inputs: Dict[str, Any],
     ) -> None:
-        if current_callback_context_data := self._current_callback_context.get():
-            if isinstance(current_callback_context_data, span.SpanData):
-                self._attach_span_to_existing_span(
-                    call_id=call_id,
-                    current_span_data=current_callback_context_data,
-                    instance=instance,
-                    inputs=inputs,
-                )
-            else:
-                self._attach_span_to_existing_trace(
-                    call_id=call_id,
-                    current_trace_data=current_callback_context_data,
-                    instance=instance,
-                    inputs=inputs,
-                )
-            return
-
-        if current_span_data := opik_context.get_current_span_data():
+        # First we check the callback's context
+        if current_span_data := self._context_storage.top_span_data():
```

Comment:
Maybe it would be better to have explicit check for not None in all these assignment if blocks.

```
if (current_span_data := self._context_storage.top_span_data()) is not None:
```

This will improve the readability of the code